File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/atomic/Circle.java
package com.greentree.commons.geometry.geom2d.atomic;

import com.greentree.commons.geometry.math.MathLine1D;
import com.greentree.commons.math.vector.AbstractVector2f;
import com.greentree.commons.math.vector.Vector2f;
import com.greentree.commons.util.collection.InfinityCollection;


public class Circle implements SimpleShape2D {
	
	public static final Circle INSTANCE = new Circle();
	
	private Circle() {
	}
	
	@Override
	public boolean isInsade(AbstractVector2f point) {
		return point.lengthSquared() <= 1;
	}
	
	@Override
	public AbstractVector2f min(AbstractVector2f point) {
		return new Vector2f(point).normalize();
	}
	
	@Override
	public MathLine1D getProjection(AbstractVector2f normal) {
		return new MathLine1D(-1, 1);
	}
	
	@Override
	public Iterable<? extends AbstractVector2f> normals() {
		return new InfinityCollection<>();
	}
	
	@Override
	public Iterable<? extends AbstractVector2f> points() {
		return new InfinityCollection<>();
	}
	
}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/atomic/Circle.kt
package com.greentree.commons.geometry.geom2d.atomic

import com.greentree.commons.math.MathLine1D
import com.greentree.commons.math.vector.AbstractVector2f
import com.greentree.commons.math.vector.Vector2f

object Circle : SimpleShape2D {

	override fun projection(normal: AbstractVector2f): MathLine1D {
		return MathLine1D(-1f, 1f)
	}

	override fun isInside(point: AbstractVector2f): Boolean {
		return point.lengthSquared() <= 1
	}

	override fun nearestPoint(point: AbstractVector2f): AbstractVector2f {
		return Vector2f(point).normalize(1)
	}

	override fun normals(): Iterable<AbstractVector2f> {
		return NormalsVector2fCollection()
	}

	override fun points(): Iterable<AbstractVector2f> {
		return NormalsVector2fCollection()
	}
}